dispatch.go

```diff
@@ -124,30 +124,33 @@ func (d *dispatch) handleCall(calls map[seqNumber]*call, c *call) {
 	seqid := d.nextSeqid()
 	c.seqid = seqid
 	calls[c.seqid] = c
-	err := d.dispatchMessage(c.ctx, MethodCall, seqid, c.method, c.arg)
-	if err != nil {
-		setResult := c.Finish(err)
-		if !setResult {
-			panic("c.Finish unexpectedly returned false")
-		}
-		return
-	}
+	errCh := d.dispatchMessage(c.ctx, MethodCall, seqid, c.method, c.arg)
 	d.log.ClientCall(seqid, c.method, c.arg)
 	go func() {
-		select {
-		case <-c.ctx.Done():
-			setResult := c.Finish(newCanceledError(c.method, seqid))
-			if !setResult {
-				// The result of the call has already
-				// been processed.
-				return
+		for {
+			select {
+			case <-c.ctx.Done():
+				d.log.Warning("Context canceled")
+				setResult := c.Finish(newCanceledError(c.method, seqid))
+				if !setResult {
+					d.log.Warning("Result of call(%d) has already been processed", seqid)
```

Comment:
would be nice to also have the method, like newCanceledError. Otherwise, it's not a very useful message